File: platform/robot/src/navigation/start_conditions.py
"""Where the robot believes it starts, without being handed a scenario file.

The starting pose is pure track geometry -- corridor centreline, biased away
from the inner block, aligned with travel -- so it can be computed from the
layout the robot *believes* it is on rather than read from metadata. That is
what lets the deployed navigator run with no scenario file at all.

This lived in ``src/simulation/scenario_builder.py``, which meant the only code
that could derive a start pose was simulation code, and the real node had to be
told one. It is re-exported from there so existing callers are unaffected.

## Why the section can be assumed but the direction cannot

**Section is free.** The robot defines its own world frame: it declares its
starting corridor to be the south one and everything else follows. If it is
physically in the east corridor, its entire map is the true map rotated 90
degrees, and since it learns corridor widths against its own labels the map
stays self-consistent. The path it drives is correct in its own frame, which is
the only frame it acts in.

**Direction is not.** Assuming south-clockwise also asserts *the inner block is
on my right*. If the robot is actually travelling the other way round the loop
the block is on its left, the assumption is wrong by a reflection rather than a
rotation, and no amount of width learning recovers it -- it steers toward the
wall it thinks is the far one. Reflection is not a symmetry of the labelled
track, so travel direction has to come from outside: a launch parameter or a
jumper, the same way the challenge mode does.
"""

from __future__ import annotations

import math
from typing import TYPE_CHECKING

from shared.config.constants import CorridorDimensions, DictKeys, TrackDimensions
from shared.domain.enums import Direction, Section

from src.config.tuning_helpers import get_tuning
from src.navigation.planning.waypoints.generation import center_bias_for_corridor
from src.navigation.race_tracker import TRAVEL_DIRS

if TYPE_CHECKING:
    from shared.config.navigation_tuning import NavigationTuning

_TRACK_CENTER = TrackDimensions.MAX_COORD / 2

CANONICAL_SECTION = Section.canonical()
"""The section a robot assumes when it has not been told which one it is in.

Any choice works -- see the module docstring -- so this is a label, not a claim
about where the robot physically is. Sourced from ``Section.canonical()``
rather than restated here so ``StartingConditions``'s own default (which
cannot import this module -- domain layer, see its docstring) stays in step
with this one automatically.
"""


def start_pose(
    section: Section,
    direction: Direction,
    widths_m: dict[str, float],
    tuning: NavigationTuning | None = None,
    center_bias_m: float | None = None,
) -> tuple[float, float, float]:
    """Spawn pose on the biased corridor centerline, aligned with travel.

    Args:
        section: Which corridor the robot starts in.
        direction: Travel direction, which sets the yaw.
        widths_m: Corridor widths by lowercase section name.
        tuning: Navigation tuning instance. Defaults to loaded defaults.
        center_bias_m: When set, EVERY corridor takes this magnitude and the
            narrow/wide split is skipped. None (the default) applies the split,
            which is what the Open planner does.

            Callers that supply it are pinning the pre-split behaviour on
            purpose. The Obstacles Challenge is the one that does: its assumed
            start is computed from the all-narrow prior at the WIDE magnitude,
            so pinning that magnitude leaves the pose bit-identical to what its
            scenarios were tuned against. Verified 2026-08-29: its failure set
            is unchanged by the split, over a matched sweep of the same seven
            test modules at the same worker count.

            Obstacles' assumed start therefore does NOT sit on its own planned
            centreline, which uses OBSTACLES_CENTER_BIAS_M; that inconsistency
            predates the split and is left alone deliberately rather than fixed
            in passing, since changing a separately swept value is its own
            measurement.

    Uses tuning: waypoints.WIDE_CENTER_BIAS_M, NARROW_CENTER_BIAS_M,
    NARROW_WIDTH_THRESHOLD_M, WIDE_CENTER_BIAS_SIDE, NARROW_CENTER_BIAS_SIDE
    """
    tuning = get_tuning(tuning)

    # Per-corridor bias, via the same helper calculate_waypoints uses, so the
    # split is never reimplemented here and cannot drift from the planner's.
    # This matters for Open specifically: the blind prior below is all-narrow,
    # so a blind Open robot plans narrow corridors centred and must assume a
    # start on that same centreline, not on the wide value it used before the
    # split existed.
    def _bias(width_m: float) -> float:
        return center_bias_for_corridor(width_m, tuning, center_bias_m)

    south_cy = widths_m["south"] / 2 + _bias(widths_m["south"])
    north_cy = TrackDimensions.MAX_COORD - widths_m["north"] / 2 - _bias(widths_m["north"])
    east_cx = TrackDimensions.MAX_COORD - widths_m["east"] / 2 - _bias(widths_m["east"])
    west_cx = widths_m["west"] / 2 + _bias(widths_m["west"])
    center = {
        Section.SOUTH: (_TRACK_CENTER, south_cy),
        Section.NORTH: (_TRACK_CENTER, north_cy),
        Section.EAST: (east_cx, _TRACK_CENTER),
        Section.WEST: (west_cx, _TRACK_CENTER),
    }[section]
    normal = TRAVEL_DIRS[(section, direction)]
    return center[0], center[1], math.atan2(normal.ny, normal.nx)


def assumed_start_conditions(
    direction: Direction,
    widths_m: dict[Section, float] | None = None,
    section: Section = CANONICAL_SECTION,
    tuning: NavigationTuning | None = None,
    center_bias_m: float | None = None,
) -> dict:
    """Starting conditions for a robot that was told nothing but the direction.

    Args:
        direction: Travel direction for the round. The one thing that cannot be
            assumed -- see the module docstring.
        widths_m: The layout the robot believes it is on. Defaults to every
            corridor narrow, the safe prior blind operation starts from. Note
            this interacts with the narrow/wide bias split: the default belief
            is NARROW, so a blind robot assumes the narrow bias until it
            measures otherwise, which is what keeps the assumed start on the
            path it will actually be given.
        section: Which corridor the robot calls its starting one.
        tuning: Navigation tuning instance. Defaults to loaded defaults.
        center_bias_m: Centreline-shift override, forwarded to
            :func:`start_pose`. Pass the same value handed to
            ``calculate_waypoints`` for this challenge.

    Returns:
        A ``starting_conditions`` mapping in scenario-metadata shape.

    Uses tuning: waypoints.WIDE_CENTER_BIAS_M, NARROW_CENTER_BIAS_M,
    NARROW_WIDTH_THRESHOLD_M, WIDE_CENTER_BIAS_SIDE, NARROW_CENTER_BIAS_SIDE
    """
    believed = widths_m or dict.fromkeys(Section, CorridorDimensions.NARROW)
    by_name = {s.value.lower(): w for s, w in believed.items()}
    sx, sy, yaw = start_pose(section, direction, by_name, tuning, center_bias_m)
    return {
        DictKeys.DIRECTION: str(direction),
        DictKeys.SECTION: section.capitalized,
        DictKeys.POSITION: {DictKeys.X: sx, DictKeys.Y: sy},
        DictKeys.YAW: yaw,
    }

File: platform/robot-go/internal/nav/startconditions/startconditions.go
// Package startconditions computes where the robot believes it starts,
// without being handed a scenario file. Ports src/navigation/start_conditions.py.
//
// The starting pose is pure track geometry -- corridor centerline, biased
// away from the inner block, aligned with travel -- so it can be computed
// from the layout the robot believes it is on rather than read from
// metadata. That is what lets the deployed navigator run with no scenario
// file at all.
//
// # Why the section can be assumed but the direction cannot
//
// Section is free: the robot defines its own world frame by declaring its
// starting corridor to be the south one, and everything else follows. If it
// is physically in the east corridor, its entire map is the true map
// rotated 90 degrees, and since it learns corridor widths against its own
// labels the map stays self-consistent. The path it drives is correct in
// its own frame, which is the only frame it acts in.
//
// Direction is not. Assuming south-clockwise also asserts the inner block
// is on the robot's right. If the robot is actually traveling the other
// way round the loop the block is on its left, the assumption is wrong by a
// REFLECTION rather than a rotation, and no amount of width learning
// recovers it -- it steers toward the wall it thinks is the far one.
// Reflection is not a symmetry of the labeled track, so travel direction
// has to come from outside: a launch parameter or a jumper, the same way
// the challenge mode does.
package startconditions

import (
	"math"

	"github.com/teamvoltimor/vtitan/platform/robot-go/internal/nav/racetracker"
	"github.com/teamvoltimor/vtitan/platform/robot-go/internal/nav/trackmodel"
	"github.com/teamvoltimor/vtitan/platform/robot-go/internal/nav/waypoints"
)

// Config is start-conditions' tuning: the track geometry StartPose needs
// plus the waypoint generator's centerline-bias split, which StartPose
// reuses rather than reimplementing so a spawn pose can never drift from
// the planner's own centerline.
type Config struct {
	// TrackMaxCoordM is the track's outer boundary
	// (profile.TrackConfig.Track.MaxCoord).
	TrackMaxCoordM float64
	// NarrowWidthM is the corridor width AssumedStartConditions believes
	// every side is, when the caller supplies no widths of its own -- the
	// safe prior blind operation starts from.
	NarrowWidthM float64
	// Waypoints parameterizes the narrow/wide centerline-bias split; see
	// waypoints.CenterBiasForCorridor.
	Waypoints waypoints.Config
}

// StartingConditions is a starting_conditions mapping in scenario-metadata
// shape, matching assumed_start_conditions' return value.
type StartingConditions struct {
	Direction trackmodel.Direction
	Section   trackmodel.Section
	X, Y, Yaw float64
}

// CanonicalSection is the section a robot assumes when it hasn't been told
// which one it is in. Any choice works -- see the package doc -- so this is
// a label, not a claim about where the robot physically is. Matches
// shared.domain.enums.Section.canonical().
const CanonicalSection = trackmodel.South

// Default track-geometry values, matching track.toml's [track]/[corridor]
// sections.
const (
	// DefaultTrackMaxCoordM matches track.toml's [track] max_coord.
	DefaultTrackMaxCoordM = 3.0
	// DefaultNarrowWidthM matches track.toml's [corridor] narrow.
	DefaultNarrowWidthM = 0.6
)

// DefaultConfig returns the Config matching the shipped TOML defaults.
func DefaultConfig() Config {
	return Config{
		TrackMaxCoordM: DefaultTrackMaxCoordM,
		NarrowWidthM:   DefaultNarrowWidthM,
		Waypoints:      waypoints.DefaultConfig(),
	}
}

// StartPose returns the spawn pose on the biased corridor centerline,
// aligned with travel, and ok=false when section/direction is not one of
// the four known corridors/travel directions.
//
// widthsM must have all four trackmodel.Section keys populated, in meters.
// centerBiasM, when non-nil, pins EVERY corridor to this magnitude and
// skips the narrow/wide split -- see waypoints.CenterBiasForCorridor for
// when a caller wants that (the Obstacles Challenge does).
func StartPose(
	section trackmodel.Section,
	direction trackmodel.Direction,
	widthsM map[trackmodel.Section]float64,
	cfg Config,
	centerBiasM *float64,
) (x, y, yaw float64, ok bool) {
	bias := func(widthM float64) float64 {
		return waypoints.CenterBiasForCorridor(widthM, cfg.Waypoints, centerBiasM)
	}

	trackCenter := cfg.TrackMaxCoordM / 2
	var cx, cy float64
	switch section {
	case trackmodel.South:
		cx, cy = trackCenter, widthsM[trackmodel.South]/2+bias(widthsM[trackmodel.South])
	case trackmodel.North:
		cx, cy = trackCenter, cfg.TrackMaxCoordM-widthsM[trackmodel.North]/2-bias(
			widthsM[trackmodel.North],
		)
	case trackmodel.East:
		cx, cy = cfg.TrackMaxCoordM-widthsM[trackmodel.East]/2-bias(
			widthsM[trackmodel.East],
		), trackCenter
	case trackmodel.West:
		cx, cy = widthsM[trackmodel.West]/2+bias(widthsM[trackmodel.West]), trackCenter
	default:
		return 0, 0, 0, false
	}

	normal, ok := racetracker.TravelNormalFor(section, direction)
	if !ok {
		return 0, 0, 0, false
	}
	return cx, cy, math.Atan2(normal.NY, normal.NX), true
}

// AssumedStartConditions returns the starting conditions for a robot that
// was told nothing but the direction, and ok=false when section/direction
// don't resolve to a pose (see StartPose).
//
// direction is the one thing that cannot be assumed -- see the package doc.
// widthsM is the layout the robot believes it is on; nil defaults to every
// corridor at cfg.NarrowWidthM, which interacts with the narrow/wide bias
// split: the default belief is narrow, so a blind robot assumes the narrow
// bias until it measures otherwise, keeping the assumed start on the path
// it will actually be given.
func AssumedStartConditions(
	direction trackmodel.Direction,
	widthsM map[trackmodel.Section]float64,
	section trackmodel.Section,
	cfg Config,
	centerBiasM *float64,
) (StartingConditions, bool) {
	believed := widthsM
	if believed == nil {
		believed = map[trackmodel.Section]float64{
			trackmodel.North: cfg.NarrowWidthM,
			trackmodel.South: cfg.NarrowWidthM,
			trackmodel.East:  cfg.NarrowWidthM,
			trackmodel.West:  cfg.NarrowWidthM,
		}
	}
	x, y, yaw, ok := StartPose(section, direction, believed, cfg, centerBiasM)
	if !ok {
		return StartingConditions{}, false
	}
	return StartingConditions{Direction: direction, Section: section, X: x, Y: y, Yaw: yaw}, true
}
